curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java

```diff
@@ -694,17 +714,46 @@ <DATA_TYPE> void processBackgroundOperation(OperationAndData<DATA_TYPE> operatio
         }
     }
 
+    private void closeOperation(OperationAndData<?> operation) {
+        if (operation.getCallback() == null) {
+            return;
+        }
+        CuratorEvent event = new CuratorEventImpl(this, CuratorEventType.CLOSING, KeeperException.Code.SESSIONEXPIRED.intValue(), null, null, operation.getContext(), null, null, null, null, null, null);
+        sendToBackgroundCallback(operation, event);
+    }
+
+    private void requeueSleepOperation(OperationAndData<?> operationAndData) {
+        operationAndData.clearSleep();
+        synchronized (closeLock)
+        {
+            if (getState() == CuratorFrameworkState.STARTED)
+            {
+                if (backgroundOperations.remove(operationAndData))
+                {
+                    // due to the internals of DelayQueue, operation must be removed/re-added so that re-sorting occurs
+                    backgroundOperations.offer(operationAndData);
+                }   // This operation has been taken over by background thread.
+                return;
+            }
+        }
+        closeOperation(operationAndData);
+    }
+
     /**
      * @param operationAndData operation entry
      * @return true if the operation was actually queued, false if not
      */
     <DATA_TYPE> boolean queueOperation(OperationAndData<DATA_TYPE> operationAndData)
     {
-        if ( getState() == CuratorFrameworkState.STARTED )
+        synchronized (closeLock)
         {
-            backgroundOperations.offer(operationAndData);
-            return true;
+            if (getState() == CuratorFrameworkState.STARTED)
```

Comment:
```suggestion
            if (isStarted())
```

nit - a bit bikeshedding

Reply:
Hmm, `isStarted` is marked as `Deprecated`. CURATOR-147(#44) targeted removing invocation of its usages, though I guess we are only deprecating the API part. Keep it as is for now ? All other usages in Curator use this shape currently.

Reply:
Thanks for your explanation! SGTM.